### PR TITLE
Socket setup and authentication

### PR DESCRIPTION
**Pull request by kevin-06-huang**
### What this PR does (required):
- Resolve issue #30
- implemented a protectSocket function in middleware
- socket now use it to authenticate
### Screenshots / Videos (front-end only):
- Post at least one screenshot to show the feature in action. **Only required in front-end work**

### Any information needed to test this feature (required):
- sign in with localhost:3000/signup then click on login button to login (assuming connected to mongodb)
- console will display user id

### Any issues with the current functionality (optional):
- List any problems where you were not able to complete all tasks on ticket
- List any problems this PR may cause for the project or other active PRs
- Post a screenshot/video of the problem, along with detailed console outputs where applicable

### CODE REVIEW

File: server/middleware/auth.js [hunk 1]
```diff
@@ -18,4 +18,22 @@ const protect = (req, res, next) => {
   }
 };
 
+const protectSocket = (socket, next) => {
+  const { cookie } = socket.handshake.headers;
+  const token = cookie.split('token=')[1];
+
+  if (!token) {
+    next(new Error('No token, authorization denied'));
```
Comment (ezeilo-su) [inline server/middleware/auth.js:26]: I think you we should `throw` at this point `throw(new Error('No token, authorization denied'));`

File: server/middleware/auth.js [hunk 2]
```diff
@@ -18,4 +18,22 @@ const protect = (req, res, next) => {
...
+  }
+
+  try {
+    const decoded = jwt.verify(token, process.env.JWT_SECRET);
+    socket.decoded = decoded;
+    next();
+  } catch (err) {
+    next(new Error('Token is not valid'));
```
Comment (ezeilo-su) [inline server/middleware/auth.js:34]: Also `throw` here

**Review by ezeilo-su — CHANGES_REQUESTED** [review]
Hey @kevin-06-huang, good progress. Let's have a look at the comments I left below.

File: server/middleware/auth.js (hunk 1 above)
Comment (kevin-06-huang) [reply server/middleware/auth.js:26]: Thank you, changed as commented in commit.

File: server/middleware/auth.js (hunk 2 above)
Comment (kevin-06-huang) [reply server/middleware/auth.js:34]: Changed in commit.

**Comment by kevin-06-huang** [comment]
All comments are addressed in new commit [d130daf](https://github.com/hatchways/team-wintergreen/pull/71#:~:text=Changed%20to%20throw%20as%20commented.)

**Review by ezeilo-su — COMMENTED** [review]
Thank you for making the requested changes!

**Review by rajivtitus — APPROVED** [review]
Great work! Additionally, it would help to keep track of all the logged in users 👍🏽

**Pull request merged**